### PR TITLE
Introduce usage of Vertex Array Objects (VAO)

### PR DESCRIPTION
**Pull request by Mugen87**
This PR introduces the usage of [Vertex Array Objects](https://blog.tojicode.com/2012/10/oesvertexarrayobject-extension.html) (VAO), a performance primitive of WebGL (see https://github.com/mrdoob/three.js/issues/8705). A new class `WebGLBindingStates` manages now the buffer bindings and pointers for a single entity. VAO can only be used via an extension, the support though is really high (96% according to http://webglstats.com/webgl/extension/OES_vertex_array_object). Besides VAOs are part of the core of WebGL2.

If the extension is not available, `three.js` renders the old way. It also uses the old render path if morph targets are present, since it's pretty complicated to manage the flexible state of morph target attributes with VAOs. Maybe we find a solution for this in the future.

The amount of changes is actually manageable. Most changes are caused by refactoring, since certain code like `setupVertexAttributes()` moved from `WebGLRenderer` to `WebGLBindingStates`. I've tested most examples and they still work, although intensive testing is important for such a change. 

Also see https://github.com/mrdoob/three.js/issues/8705#issuecomment-296177105

### CODE REVIEW

**Comment by Mugen87** [comment]
There are of course some open points. I'm not sure how to safely dispose individual `VAO`s if they are not necessary anymore. Besides, if users add buffer attributes after the first `bindingStates.setup()`, they won't see any effect since the `VAO` is generate once and then reused.

**Comment by takahirox** [comment]
I've been working on VAO, too.

https://github.com/takahirox/three.js/tree/VAOImpleTrial

> Besides, if users add buffer attributes after the first bindingStates.setup(), they won't see any effect since the VAO is generate once and then reused.

Yup, that's the problem. In my branch, I'm trying to add `geometry.version`

https://github.com/takahirox/three.js/blob/VAOImpleTrial/src/core/BufferGeometry.js#L75
https://github.com/takahirox/three.js/blob/VAOImpleTrial/src/renderers/webgl/WebGLVertexArrayObjects.js#L81

`.version` will be count up on `BufferGeometry`'s attribute handling methods and `setupVertexAttributes()` will be called in renderer's `render()` if it's updated.

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -836,136 +823,9 @@ function WebGLRenderer( parameters ) {
...
-					}
-
-				}
-
-			}
-
-		}
-
-		state.disableUnusedAttributes();
+		bindingStates.resetBindingState();
```
Comment (takahirox) [inline src/renderers/WebGLRenderer.js]: (IIRC) I saw performance degradation by resetting VAO in the end of each `.renderBufferDirect()` call in some examples. I'll share the experiment later.

**Comment by Mugen87** [comment]
> I'm trying geometry.version

Yeah, that could work. We just need one way to distinct different attribute setups.

> thinking clean up VAOs in renderer's dispose() so far.

Yeah, I've implemented it like that. But we introduce a memory leak if we have no other option.

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (Mugen87) [reply src/renderers/WebGLRenderer.js]: This line was actually necessary to ensure that other render calls (e.g. by `WebGLSpriteRenderer`) don't render with an invalid binding state. It would actually be surprising if this call has a negative impact on performance since I've never saw it differently used.

Comment (takahirox) [reply src/renderers/WebGLRenderer.js]: > This line was actually necessary to ensure that other render calls (e.g. by WebGLSpriteRenderer) don't render with an invalid binding state.

I know. I first tried the same approach but I saw a negative impact and started to think the way to remove that line. Removing that line makes the code a bit complicated. This's one of the reasons why I'm late to share my work.

**Comment by Mugen87** [comment]
Okay, if the configuration of a `BufferGeometry` is done via its designated methods (`setIndex()`, `addAttribute()` and `removeAttribute()`, the usage of `version` should be robust.

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (takahirox) [reply src/renderers/WebGLRenderer.js]: Remembering... One of the cases where I saw performance degradation is rendering many `Mesh`es which share the same geometry and material, like `performance` example.
In that case, ideally we don't need to reset for each `Mesh`. 

But in this implementation it calls `extension.bindVertexArrayOES( vao );` `extension.bindVertexArrayOES( null );` in each `.renderBufferDirect()`. These additional calls (compared to the current implement) seem to lead a negative impact.

IMO, I don't wanna accept this performance degradation. VAO is for improve the performance. Having a negative impact even on existing example doesn't make sense.

Comment (takahirox) [reply src/renderers/WebGLRenderer.js]: So in my branch, I save the current VAO and check if we need to switch VAO by calling `.bindVertexArrayOES()`. 

But, as you know, no calling `extension.bindVertexArrayOES( null );` means VAO can be polluted somewhere. I found `WebGLAttribute` `create/updateBuffer` pollutes index(`gl.ELEMENT_ARRAY_BUFFER`).

So I'm trying a workaround

https://github.com/takahirox/three.js/blob/VAOImpleTrial/src/renderers/webgl/WebGLAttributes.js#L16
https://github.com/takahirox/three.js/blob/VAOImpleTrial/src/renderers/webgl/WebGLVertexArrayObjects.js#L93

but I know it isn't good design. I'm also thinking better solution. Perhaps we need further refactoring?

Comment (Mugen87) [reply src/renderers/WebGLRenderer.js]: @kenrussell To sum this up: We are not sure about the overhead of a single `extension.bindVertexArrayOES( null );`. Is it bad if we do this for each render item in order to ensure a consistent binding state?

Comment (takahirox) [reply src/renderers/WebGLRenderer.js]: Not single `extension.bindVertexArrayOES( null );` overhead, but the set of `extension.bindVertexArrayOES( vao );` and `extension.bindVertexArrayOES( null );` overhead in the beginning and the end of rendering an item.

Comment (Mugen87) [reply src/renderers/WebGLRenderer.js]: Right, we need to perform both calls for each item.

Comment (takahirox) [reply src/renderers/WebGLRenderer.js]: > I think it's important to have a better average performance over many examples.

I can't perfectly agree. Sharing Geometry and Material should be one of the good optimizations on memory and performance. With such an optimization, seeing a negative performance impact is bad. We should investigate more how to achieve **both** less changes and less(or no) performance degradation.

Comment (Mugen87) [reply src/renderers/WebGLRenderer.js]: > We should investigate more how to achieve both less changes and less(or no) performance degradation.

That's fine.

Comment (takahirox) [reply src/renderers/WebGLRenderer.js]: In real-time 3D field, 60 (or more in VR) fps is a performance threshold. In average cases, probably the apps are already light enough and run on 60 fps without VAO. In practical most user don't see any good impact by introducing VAO. But they could see the performance degradation on heavy but well optimized apps.

IMO we should go more carefully with VAO. Targeting only average might not be good for VAO.

**Comment by Mugen87** [comment]
Okay, the last commit restores the previous behavior. The implementation now uses `geometryProgram` and only performs `extension.bindVertexArrayOES( bindingState );` if the identifier is different. `bindingStates.resetBindingState();` is now only performed once after the opaque and transparent pass. Subsequent operations are thus not affected by inconsistent binding state.

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (Mugen87) [reply src/renderers/WebGLRenderer.js]: I just wanted to point out that we need to find a compromise between performance a the complexity of the implementation.

**Comment by Mugen87** [comment]
I'm closing this PR since I think we need a different approach in order to address @takahirox's feedback. For a first go, it was quite insightful 😊.

tl;dr: We need a more sophisticated approach in order to minimize calls to `bindVertexArrayOES`. In any event, it would be still useful to know about the performance implications of s single `bindVertexArrayOES` call.

**Comment by Mugen87** [comment]
@takahirox Many thanks for this seesion 😉

**Comment by takahirox** [comment]
@Mugen87 My pleasure and thanks, too! Can we work for VAO together? It'd be more efficient.

**Comment by Mugen87** [comment]
Of course we can 😊. TBH, I did not know that you are also working on this. It would have saved some time if I studied your branch first.

Anyway, for now I'm out of ideas and need to reconsider everything. Using VAOs leads to new problems within the engine I was not aware of...

**Comment by kenrussell** [comment]
Sorry, coming late to the game.

In general, one call to bindVertexArray (using the WebGL 2.0 naming) should be cheaper than multiple calls to bindBuffer + vertexAttribPointer to set up all of the attributes for the program. Binds are still expensive though, so yes, it's important to try to minimize them.

To measure this best, try a scene that has a lot of objects and which might not be rendered with the most efficient technique (like instancing). Ideally, one which doesn't reach 60 FPS on your machine because it's GPU bound. Then compare the bindVertexArray code path to the previous one.

If you see the bindVertexArray code path being slower even after you've eliminated useless bindVertexArray(null) calls, then please package up the before-and-after test cases and file a bug on crbug.com, attach the test case, and send me the bug ID. We'll profile the WebGL implementation and driver and see what's happening.

**Comment by Mugen87** [comment]
@kenrussell Okay, many thanks for your feedback 👍

**Comment by Mugen87** [comment]
@takahirox I think we should study how babylon.js uses VAOs. We should find everything important in [babylon.engine.ts](https://github.com/BabylonJS/Babylon.js/blob/master/src/Engine/babylon.engine.ts). Especially methods like `recordVertexArrayObject()` are interesting. The engine maintains a boolean `vaoRecordInProgress` that indicates the creation of a VAO. If they call methods to configure attributes or the element array buffer, they check `vaoRecordInProgress`. If no record is in process, they unbind the current VAO to perform the actual `gl. bindBuffer`. That's similar what you suggested yesterday.

https://github.com/BabylonJS/Babylon.js/blob/4a222735e01883bd30af3cf18100eabc734107bc/src/Engine/babylon.engine.ts#L3030-L3045

https://github.com/BabylonJS/Babylon.js/blob/4a222735e01883bd30af3cf18100eabc734107bc/src/Engine/babylon.engine.ts#L2920-L2925

**Comment by takahirox** [comment]
@Mugen87 Thanks. Let me check the babylon code later.

**Comment by takahirox** [comment]
@Mugen87 

If I understand the babylon code correctly, yes I think we need such a check.

BTW, if I'm right, I don't think the check is necessary for attributes because `.bindBuffer` for attributes doesn't pollute VAO.

https://www.khronos.org/opengl/wiki/Vertex_Specification#Vertex_Buffer_Object

**Comment by Mugen87** [comment]
> A call to glBindBuffer to set the GL_ARRAY_BUFFER binding is NOT modifying the current VAO's state!

Interesting. According to the following [post](https://stackoverflow.com/a/20910235/5250847) that seems also true for `bufferData`.

> It's also worth noting that you can always update the data in a VBO using glBufferData or similar calls, and that operation will have no effects on the binding contained with the VAO associated with that buffer.

In this case, changes to `WebGLAttributes` should actually no be necessary?

**Comment by takahirox** [comment]
Yup, if I'm right `.bufferData` doesn't affect VAO status. But `.bindBuffer` for indexBuffer (`.gl.ELEMENT_ARRAY_BUFFER`) pollutes VAO.

In my repository, I check if `bufferType` is `gl.ELEMENT_ARRAY_BUFFER` for `.bindBuffer` not to pollute VAO.

https://github.com/takahirox/three.js/blob/VAOImpleTrial/src/renderers/webgl/WebGLAttributes.js#L16
https://github.com/takahirox/three.js/blob/VAOImpleTrial/src/renderers/webgl/WebGLVertexArrayObjects.js#L95
